### ISSUE TITLE
Multi-threaded compilation

### ISSUE THREAD

**Issue by tiiago11** [KhronosGroup/glslang#2438]
Hello,

I'm looking for documentation/examples regarding multi-threaded shader compilation using the C++ API in in order to compile from GLSL -> SPIRV. In the standalone app there is a use case with the legacy API only. 

From early tests I didn't have issues, but I wonder what is the correct/intended way to achieve this. And most importantly, whether it is safe to use the new C++ API in such a multi-threaded way. 

e.g., something like
```
parallel_for(0, incomingShaderCode.size(), [](int index)
{
     ... do setup of the shader (single shader)
     TShader* shader = new TShader;
     shader->setStringsWithLengthsAndNames(incomingShaderCode[index].code, ...)
     shader->parse(...);

      ... create the program and generate SPIRV
     glslang::TProgram& program = *new glslang::TProgram;
     program.addShader(shader);
     glslang::GlslangToSpv(program.getIntermediate(), ...)

     ... cleanup
} 
````

Thank you in advance.

**Comment by johnkslang** [comment]
Yes, it should all be thread safe, and independent of which API is used to request the work.

I think the point about StandAlone is that the part it runs in parallel just includes GLSL/HLSL -> AST, and is not yet including AST -> SPIR-V, so that's something that could be added to StandAlone.cpp.  It would be a bit of a new thing to do, as AST -> SPIR-V requires linking first.  The split between linking and non-linking was just artificial, testing the diagonals of this matrix instead of all four quadrants:

```
                   just AST          linking & SPV
single thread                              X
multi threaded        X
```

But, I suspect there are other drivers of the API (possibly shaderc, and at least one engine?) that cover multi-threaded SPV generation.

**Comment by tiiago11** [comment]
Thank you for the quick reply!

Yeah, since there were no tests I was wondering if it was recommended/safe, but now it's clear.
I guess some tests would be nice to have, also to detect possible regressions.

**Comment by johnkslang** [comment]
Agreed. Will morph this issue into that.
